Core/src/IfcGeometryConverter/FaceConverter.h

```diff
@@ -801,59 +790,73 @@ namespace OpenInfraPlatform {
+					const carve::math::Matrix& pos) const noexcept(true)
+				{
+					const std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> faceBounds = swapOuterBoundaryToFront(ifcFaceBounds);
+
+					std::vector<std::vector<carve::geom::vector<3>>> faceBoundLoops;
+					faceBoundLoops.resize(faceBounds.size());
+					for (size_t i = 0; i < faceBounds.size(); i++)
+					{
+						// get vertices of one bound loop into faceBoundLoops[i]
+						faceBoundLoops[i] = convertIfcFaceBound(faceBounds[i], pos);
+					}
+
+					return faceBoundLoops;
+				}
+
+				std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> swapOuterBoundaryToFront(
+					std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>> ifcFaceBounds) const throw(...)
 				{
 					// As carve expects outer boundary of face to be at first index, outer boundary has to be moved to index 0 and inner boundary has index >= 1
 					for (auto it = ifcFaceBounds.begin(); it != ifcFaceBounds.end(); it++)
 					{
 						if (it->isOfType<typename IfcEntityTypesT::IfcFaceOuterBound>())
 						{
-							// swap element 'it' with first element
+							// swap element 'it' with first element;
 							std::rotate(ifcFaceBounds.begin(), it, it + 1);
 
 							// according to ifc-documentation, only one boundary (= loop) can be outer boundary:
 							// "One loop is optionally distinguished as the outer loop of the face."
 							// (https://standards.buildingsmart.org/IFC/DEV/IFC4_3/RC3/HTML/link/ifcface.htm)
 							// thus, break the for-loop after THE outer boundary was found
-							break;
+							return ifcFaceBounds;
 						}
 					}
-
-					faceBoundLoops.resize(ifcFaceBounds.size());
-					for (size_t i = 0; i < ifcFaceBounds.size(); i++)
-					{
-						// get vertices of one bound loop into faceBoundLoops[i]
-						convertIfcFaceBound(ifcFaceBounds[i], pos, faceBoundLoops[i]);
-					}
 				}
 
-				void convertIfcFaceBound(
+				std::vector<carve::geom::vector<3>> convertIfcFaceBound(
 					const EXPRESSReference<typename IfcEntityTypesT::IfcFaceBound>& bound,
-					const carve::math::Matrix& pos,
-					std::vector<carve::geom::vector<3>>& loopVertices3D) const noexcept(false)
+					const carve::math::Matrix& pos) const noexcept(false)
 				{
+					if (bound.expired()) {
+						throw oip::ReferenceExpiredException(bound);
+					}
+
 					//	IfcLoop   has subtypes   IfcEdgeLoop, IfcPolyLoop, IfcVertexLoop
 					const EXPRESSReference<typename IfcEntityTypesT::IfcLoop>& loop = bound->Bound;
 					bool polyOrientation = bound->Orientation;
 
-					if (!loop) {
-						throw oip::InconsistentModellingException(bound, " No valid loop");
+					if (loop.expired()) {
+						throw oip::ReferenceExpiredException(loop);
 					}
 
+					// declare target variable
+					std::vector<carve::geom::vector<3>> loopVertices3D;
+
 					// Collect all vertices of the current loop
 					curveConverter->convertIfcLoop(loop, loopVertices3D);

```

Comment:
perhaps change the return value of the function?

Reply:
Done in commit 8cc1112.
The function `convertIfcEdgeLoop` will receive some futher attention within this pull request.